test/e2e/clusterctl_upgrade.go

```diff
@@ -0,0 +1,300 @@
+						n++
+					}
+				}
+			}
+			return n, nil
+		}, input.E2EConfig.GetIntervals(specName, "wait-worker-nodes")...).Should(Equal(*controlPlaneMachineCount + *workerMachineCount))
+
+		By("THE MANAGEMENT CLUSTER WITH OLDER VERSION OF PROVIDERS WORKS!")
+
+		By("Upgrading providers to the latest version available")
+		clusterctl.UpgradeManagementClusterAndWait(ctx, clusterctl.UpgradeManagementClusterAndWaitInput{
+			ClusterctlConfigPath: input.ClusterctlConfigPath,
+			ClusterProxy:         managementClusterProxy,
+			ManagementGroup:      fmt.Sprintf("capi-system/%s", config.ClusterAPIProviderName),
+			Contract:             clusterv1.GroupVersion.Version,
+			LogFolder:            filepath.Join(input.ArtifactFolder, "clusters", cluster.Name),
+		}, input.E2EConfig.GetIntervals(specName, "wait-controllers")...)
+
+		By("THE MANAGEMENT CLUSTER WAS SUCCESSFULLY UPGRADED!")
+
+		// After upgrading we are sure the version is the latest version of the API,
+		// so it is possible to use the standard helpers
+
+		testMachineDeployments := framework.GetMachineDeploymentsByCluster(ctx, framework.GetMachineDeploymentsByClusterInput{
+			Lister:      managementClusterProxy.GetClient(),
+			ClusterName: clusterName,
+			Namespace:   testNamespace.Name,
+		})
+
+		framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+			ClusterProxy:              managementClusterProxy,
+			Cluster:                   &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Namespace: testNamespace.Name}},
+			MachineDeployment:         testMachineDeployments[0],
+			Replicas:                  2,
+			WaitForMachineDeployments: input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
+
+		By("THE UPGRADED MANAGEMENT CLUSTER WORKS!")
+
+		By("PASSED!")
+	})
+
+	AfterEach(func() {
+		if testNamespace != nil {
+			// Dump all the logs from the workload cluster before deleting them.
+			managementClusterProxy.CollectWorkloadClusterLogs(ctx, testNamespace.Name, clusterName, filepath.Join(input.ArtifactFolder, "clusters", clusterName, "machines"))
+
+			framework.DumpAllResources(ctx, framework.DumpAllResourcesInput{
+				Lister:    managementClusterProxy.GetClient(),
+				Namespace: managementClusterNamespace.Name,
+				LogPath:   filepath.Join(input.ArtifactFolder, "clusters", managementClusterResources.Cluster.Name, "resources"),
+			})
+
+			if !input.SkipCleanup {
+				Byf("Deleting cluster %s/%s", testNamespace.Name, clusterName)
+				framework.DeleteAllClustersAndWait(ctx, framework.DeleteAllClustersAndWaitInput{
+					Client:    managementClusterProxy.GetClient(),
+					Namespace: testNamespace.Name,
+				}, input.E2EConfig.GetIntervals(specName, "wait-delete-cluster")...)
+
+				Byf("Deleting namespace used for hosting the %q test", specName)
+				framework.DeleteNamespace(ctx, framework.DeleteNamespaceInput{
+					Deleter: managementClusterProxy.GetClient(),
+					Name:    testNamespace.Name,
+				})
+			}
```

Comment:
Should we also have to delete the management cluster?

Reply:
This is already implemented in `dumpSpecResourcesAndCleanup`.